cranelift/jit/src/memory/arena.rs

```diff
@@ -0,0 +1,215 @@
+use std::io;
+use std::mem::ManuallyDrop;
+use std::ptr;
+
+use cranelift_module::ModuleResult;
+
+use super::{BranchProtection, JITMemoryProvider};
+
+fn align_up(addr: usize, align: usize) -> usize {
+    debug_assert!(align.is_power_of_two());
+    (addr + align - 1) & !(align - 1)
+}
+
+#[derive(Debug)]
+struct Segment {
+    ptr: *mut u8,
+    len: usize,
+    position: usize,
+    target_prot: region::Protection,
+    finalized: bool,
+}
+
+impl Segment {
+    fn new(ptr: *mut u8, len: usize, target_prot: region::Protection) -> Self {
+        let mut segment = Segment {
+            ptr,
+            len,
+            target_prot,
+            position: 0,
+            finalized: false,
+        };
+        // set setgment to read-write for initialization
+        segment.set_rw();
+        segment
+    }
+
+    fn set_rw(&mut self) {
+        unsafe {
+            region::protect(self.ptr, self.len, region::Protection::READ_WRITE)
+                .expect("unable to change memory protection for jit memory segment");
+        }
+    }
+
+    fn finalize(&mut self, branch_protection: BranchProtection) {
+        if self.finalized {
+            return;
+        }
+
+        if self.target_prot == region::Protection::READ_EXECUTE {
+            super::set_readable_and_executable(self.ptr, self.len, branch_protection)
+                .expect("unable to set memory protection for jit memory segment");
+        } else {
+            unsafe {
+                region::protect(self.ptr, self.len, self.target_prot)
+                    .expect("unable to change memory protection for jit memory segment");
+            }
+        }
+        self.finalized = true;
```

Comment:
After finalization, `has_space_for` should unconditionally return `false` at least when this is an executable segment as any attempt to write to it will fail.